sage_code/strong_uniform_bounds.py

```diff
@@ -0,0 +1,184 @@
+"""strong_uniform_bounds.py
+
+    Computes the strong uniform bounds from the second paper.
+
+    ====================================================================
+
+    This file is part of Isogeny Primes.
+
+    Copyright (C) 2023 Barinder S. Banwait and Maarten Derickx
+
+    Isogeny Primes is free software: you can redistribute it and/or
+    modify it under the terms of the GNU General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+    The authors can be reached at: barinder.s.banwait@gmail.com and
+    maarten@mderickx.nl.
+
+    ====================================================================
+
+"""
+
+from sage.all import Partitions, divisors, matrix, GF, lcm, gcd, prime_range, prod, Integer
+from itertools import product
+import logging
+from .common_utils import get_weil_polys, R, x
+from .pil_integers import collapse_tuple
+from .type_one_primes import cached_bad_formal_immersion_data
+
+logger = logging.getLogger(__name__)
+
+def is_non_increasing(t, part):
+
+    for j in range(len(t)-1):
+        if part[j] == part[j+1]:
+            if t[j] < t[j+1]:
+                return False
+    return True
+
+
+def get_beta_mats_with_pow(F, pow=1):
+
+    frob_polys = get_weil_polys(F)
+    output = [matrix.companion(f) ** pow for f in frob_polys]
+
+    # We also need to add \pm 1, \pm q
+
+    f1 = (x - 1) * (x + 1)
+    output.append(matrix.companion(f1))
+
+    q = F.cardinality()
+    fq = (x - q) * (x + q)
+    output.append(matrix.companion(fq))
```

Comment:
oh, also check that this part, adding \pm 1, \pm q into the S set is being done correctly ...

Reply:
You could also add the companion matrices of (x-1), (x+1), (x-q) and (x+q) separately. But I'm not even sure wether this will actually be an improvement over the code as it is now. This second part of the comment is just for your information no need to actually split the companion matrices.